eng/tools/mgmtreport/main.go

```diff
@@ -0,0 +1,504 @@
+					if err != nil {
+						return nil, err
+					}
+
+					// mock test
+					err = mockTest(testClient, &mInfo, buildId)
+					if err != nil {
+						return nil, err
+					}
+
+					// live test
+					err = liveTest(buildClient, &mInfo, buildId)
+					if err != nil {
+						return nil, err
+					}
+
+					moduleName := fmt.Sprintf("%s/%s", dir.Name(), arm.Name())
+					mgmtReport[moduleName] = mInfo
+				}
+			}
+		}
+	}
+
+	return mgmtReport, nil
+}
+
+func readAutorestMD(path string) (string, string, error) {
+	var (
+		tag           string
+		readmeLink    string
+		moduleVersion string
+		multiModule   string
+	)
+
+	b, err := os.ReadFile(path)
+	if err != nil {
+		return "", "", err
+	}
+
+	lines := strings.Split(string(b), "\n")
+	for _, line := range lines {
+		if strings.Contains(line, "tag:") {
+			tag = line[len("tag:"):]
+		} else if strings.Contains(line, "readme.md") {
+			readmeLink = line[len("- "):]
+		} else if strings.Contains(line, "module-version: ") {
+			moduleVersion = line[len("module-version: "):]
+		} else if strings.Contains(line, "package-") && strings.Contains(line, ": true") {
+			multiModule = line[:len(line)-len(": true")]
+		}
+	}
+
+	if tag == "" && readmeLink != "" {
+		readmeLink = strings.ReplaceAll(readmeLink, "https://github.com", "https://raw.githubusercontent.com")
+		readmeLink = strings.ReplaceAll(readmeLink, "/blob", "")
+		resp, err := http.Get(readmeLink)
+		if err != nil {
+			return "", "", err
+		}
+
+		readmeBody, err := io.ReadAll(resp.Body)
+		if err != nil {
+			return "", "", err
+		}
+
+		if multiModule != "" {
+			indexMultiModule := bytes.Index(readmeBody, []byte(fmt.Sprintf("``` yaml $(%s)", multiModule)))
+			readmeBody = readmeBody[indexMultiModule:]
+		}
+
+		for _, line := range strings.Split(string(readmeBody), "\n") {
+			if strings.HasPrefix(line, "tag: ") {
+				tag = line[len("tag: "):]
+				break
+			}
+		}
+	}
+
+	return tag, moduleVersion, nil
+}
+
+type mgmtInfo struct {
+	tag              string
+	version          string
+	mockTestPass     int
+	mockTestTotal    int
+	CoveredLines     int
+	CoverableLines   int
+	liveTestCoverage string
+}
+
+func codeCoverage(pipelineClient pipelines.Client, azureDevopsClient *azuredevops.Client, info *mgmtInfo, pid int) (*int, error) {
+	listRuns, err := pipelineClient.ListRuns(context.Background(), pipelines.ListRunsArgs{Project: &projectName, PipelineId: &pid})
+	if err != nil && len(*listRuns) > 0 {
+		return nil, err
+	}
+
+	var buildId *int
+	for i := 0; i < 5 && i < len(*listRuns); i++ {
+		buildId = (*listRuns)[i].Id
+
+		// code coverage
+		buildCodeCoverage, err := getBuildCodeCoverage(azureDevopsClient, projectName, *buildId)
+		if err != nil {
+			return nil, err
+		}
+
+		for _, coverage := range *buildCodeCoverage.CoverageData {
+			if len(*coverage.CoverageStats) > 0 {
+				info.CoveredLines = *(*coverage.CoverageStats)[0].Covered
+				info.CoverableLines = *(*coverage.CoverageStats)[0].Total
+				break
+			}
+		}
```

Comment:
I think the parsing of coverage result should be in the `getBuildCodeCoverage` function.

Reply:
This part is not easy to move into GetBuildCodecoverage. `info` is a parameter of GetCodeCoverage